File: actions_cpp/src/move_robot_server.cpp
#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "my_robot_interfaces/action/move_robot.hpp"

using MoveRobot = my_robot_interfaces::action::MoveRobot;
using MoveRobotGoalHandle = rclcpp_action::ServerGoalHandle<MoveRobot>;
using namespace std::placeholders;

//V0 - Server
class MoveRobotServerNode : public rclcpp::Node
{
public:
    MoveRobotServerNode() : Node("count_until_server"), robot_position_(50)   
    {
        cb_group_ = this->create_callback_group(rclcpp::CallbackGroupType::Reentrant);
        move_robot_server_ = rclcpp_action::create_server<MoveRobot>(
            this,
            "move_robot",
            std::bind(&MoveRobotServerNode::goal_callback, this, _1, _2),
            std::bind(&MoveRobotServerNode::cancel_callback, this, _1),
            std::bind(&MoveRobotServerNode::handle_accepted_callback, this, _1),
            rcl_action_server_get_default_options(),
            cb_group_
        );

        RCLCPP_INFO(this->get_logger(), "Action server has been started...");
        RCLCPP_INFO(this->get_logger(), "Robot position: %d", robot_position_);
    }
private:

    rclcpp_action::GoalResponse goal_callback(
        const rclcpp_action::GoalUUID &uuid,
         std::shared_ptr<const MoveRobot::Goal> goal)
         {
            (void) uuid;
            RCLCPP_INFO(this->get_logger(), "Received a goal");            

            if((goal->position > 100 || goal->position < 0) || (goal->velocity <=0)){
                RCLCPP_WARN(this->get_logger(), "Rejecting the goal");
                return rclcpp_action::GoalResponse::REJECT;
            }            

            //Policy3: Preempt the existing goal when receiving new valid goal
            //Lock del thread   -> googlear scope en c++ lock y mutex
            {
                std::lock_guard<std::mutex> lock(mutex_);
                if(goal_handle_){
                    if(goal_handle_->is_active()){
                        preemted_goal_id_ = goal_handle_->get_goal_id();
                        RCLCPP_WARN(this->get_logger(), "A new goal is received, Abort current goal & accept new goal...");
                    }
                }
            }  

            RCLCPP_INFO(this->get_logger(), "Accepting the goal");
            return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
         }

    rclcpp_action::CancelResponse cancel_callback(
        const std::shared_ptr<MoveRobotGoalHandle> goal_handle)
         {
            RCLCPP_INFO(this->get_logger(), "Received cancel request");
            (void)goal_handle;
            return rclcpp_action::CancelResponse::ACCEPT;
         }

    void handle_accepted_callback(const std::shared_ptr<MoveRobotGoalHandle> goal_handle)
    {
        RCLCPP_INFO(this->get_logger(), "Executing the goal...");
        execute_goal(goal_handle);
    }

    void execute_goal(const std::shared_ptr<MoveRobotGoalHandle> goal_handle)
    {
        //Lock del thread
        {
            std::lock_guard<std::mutex> lock(mutex_);
            this->goal_handle_ = goal_handle;
        }       
        
        //Initial position

        //Get reuqest from goal
        int goal_position = goal_handle->get_goal()->position;
        int velocity = goal_handle->get_goal()->velocity;

        //Execute the action
        rclcpp::Rate loop_rate(1.0);

        auto feedback = std::make_shared<MoveRobot::Feedback>();
        //set the final state an return result
        auto result = std::make_shared<MoveRobot::Result>();
        //****Logica del robot*****
        while(rclcpp::ok){
            {
                std::lock_guard<std::mutex> lock(mutex_);
                if(goal_handle->get_goal_id()== preemted_goal_id_)
                {
                    result->position = robot_position_;
                    result->message = "Aborted-> Preemted by another goal";
                    goal_handle->abort(result);
                    return;
                }                
            }            
            if(goal_handle->is_canceling()){
                result->position = robot_position_;
                result->message = "Canceled";
                goal_handle->canceled(result);
                return;
            }

            int diff = goal_position - robot_position_;

            if(diff==0)
            {
                result->position = robot_position_;
                result->message = "Finished";
                goal_handle->succeed(result);
                return;
            }
            if(diff>0)
            {
                if(diff>=velocity)
                {
                    robot_position_ += velocity;
                }
                else
                {
                    robot_position_ += diff;
                }
            }
            else
            {
                if(abs(diff)>=velocity)
                {
                    robot_position_ -= velocity;
                }
                else
                {
                    robot_position_ -= abs(diff);
                }
            }
            
            
            RCLCPP_INFO(this->get_logger(), "Robot positiiom: %d", robot_position_);
            feedback->current_position = robot_position_;
            goal_handle->publish_feedback(feedback);
            loop_rate.sleep();
        }           

        
        result->position = robot_position_;
        result->message = "Finished";
        //goal_handle->abort(result);
        goal_handle->succeed(result);
        
    }

    rclcpp_action::Server<MoveRobot>::SharedPtr move_robot_server_;

    rclcpp::CallbackGroup::SharedPtr cb_group_;

    std::shared_ptr<MoveRobotGoalHandle> goal_handle_;

    std::mutex mutex_;

    rclcpp_action::GoalUUID preemted_goal_id_;

    // 2. Aquí es el mejor lugar para colocar la variable
    // Representa el estado interno del robot, por lo que debe ser privada.
    int robot_position_; 
    
    // Si usas múltiples hilos (MultiThreadedExecutor), 
    // considera usar std::atomic para evitar condiciones de carrera:
    // std::atomic<int> robot_position_;

};

int main(int argsc, char **argv){
    rclcpp::init(argsc, argv);
    auto node = std::make_shared<MoveRobotServerNode>();
    rclcpp::executors::MultiThreadedExecutor executor;
    executor.add_node(node);
    executor.spin();
    rclcpp::shutdown();
    return 0;
}

File: lifecycle_py/lifecycle_py/move_robot_server.py
#!/usr/bin/env python3
import rclpy
import time
import threading
from rclpy.lifecycle import LifecycleNode
from rclpy.lifecycle.node import LifecycleState, TransitionCallbackReturn
from rclpy.action import ActionServer
from rclpy.action.server import ServerGoalHandle, GoalResponse, CancelResponse
from my_robot_interfaces.action import MoveRobot
from rclpy.executors import MultiThreadedExecutor
from rclpy.callback_groups import ReentrantCallbackGroup


class MoveRobotServerNode(LifecycleNode): 
    def __init__(self):
        super().__init__("move_robot_server")
        self.get_logger().info("In Constructor")
        self.goal_lock_ = threading.Lock()
        self.goal_handle_: ServerGoalHandle = None
        self.robot_position_ = 50     
        self.server_activated_ = False
        self.get_logger().info("Robot position: " + str(self.robot_position_))

    ##LIFECYCLE METHODS

    def on_configure(self, previous_state: LifecycleState):
        # Recibir un parametro para modificar el Server Name para multiples servers
        self.declare_parameter("robot_name", rclpy.Parameter.Type.STRING)
        self.robot_name_ = self.get_parameter("robot_name").value
        self.get_logger().info("IN on_configure.")
        # Configurar el server
        self.move_robot_server_ = ActionServer(
            self,
            MoveRobot,
            "move_robot"+ self.robot_name_,
            goal_callback=self.goal_callback,
            cancel_callback=self.cancel_callback,
            execute_callback=self.execute_callback,
            callback_group=ReentrantCallbackGroup())
        self.get_logger().info("Action server has been started")    
        return TransitionCallbackReturn.SUCCESS
    
    def on_cleanup(self, previous_state: LifecycleState):
        self.get_logger().info("IN on_cleanup.")
        self.undeclare_parameter("robot_name")
        self.robot_name_ = ""
        self.move_robot_server_.destroy()
        return TransitionCallbackReturn.SUCCESS
    
    def on_activate(self, previous_state: LifecycleState):
        self.get_logger().info("IN on_activate.")
        self.server_activated_  = True
        return super().on_activate(previous_state)
    
    def on_deactivate(self, previous_state: LifecycleState):
        self.get_logger().info("IN on_deactivate.")
        self.server_activated_= False
        with self.goal_lock_:
            if self.goal_handle_ is not None and self.goal_handle_.is_active:
                self.goal_handle_.abort()
        return super().on_deactivate(previous_state)
    
    def on_shutdown(self, previous_state: LifecycleState):
        self.get_logger().info("IN on_shutdown.")
        self.undeclare_parameter("robot_name")
        self.robot_name_ = ""
        self.move_robot_server_.destroy()
        return TransitionCallbackReturn.SUCCESS
    
    # def on_error(self, previous_state: LifecycleState):
    #     self.get_logger().info("IN on_error.")        
    #     return TransitionCallbackReturn.SUCCESS


    ## ACTIONS Methods
    def goal_callback(self, goal_request: MoveRobot.Goal):
        self.get_logger().info("Received a new goal")

        if not self.server_activated_:
            self.get_logger().warn("Not activated yet..")
            return GoalResponse.REJECT

        if goal_request.position not in range(0, 100) or goal_request.velocity <= 0:
            self.get_logger().warn("Invalid position/velocity, reject goal")
            return GoalResponse.REJECT
        
        # New goal is valid, abort previous goal and accept new goal
        with self.goal_lock_:
            if self.goal_handle_ is not None and self.goal_handle_.is_active:
                self.goal_handle_.abort()
        
        self.get_logger().info("Accept goal")
        return GoalResponse.ACCEPT
    
    def cancel_callback(self, goal_handle: ServerGoalHandle):
        self.get_logger().info("Received a cancel request")
        return CancelResponse.ACCEPT

    def execute_callback(self, goal_handle: ServerGoalHandle):
        with self.goal_lock_:
            self.goal_handle_ = goal_handle

        goal_position = goal_handle.request.position
        velocity = goal_handle.request.velocity

        result = MoveRobot.Result()
        feedback = MoveRobot.Feedback()

        self.get_logger().info("Execute goal")
        while rclpy.ok():
            if not goal_handle.is_active:
                result.position = self.robot_position_
                result.message = "Preempted by another goal, or node deactivated"
                return result
            
            if goal_handle.is_cancel_requested:
                result.position = self.robot_position_
                if goal_position == self.robot_position_:
                    result.message = "Success after cancel request"
                    goal_handle.succeed()
                else:
                    result.message = "Canceled"
                    goal_handle.canceled()
                return result

            diff = goal_position - self.robot_position_

            if diff == 0:
                result.position = self.robot_position_
                result.message = "Success"
                goal_handle.succeed()
                return result
            elif diff > 0:
                if diff >= velocity:
                    self.robot_position_ += velocity
                else:
                    self.robot_position_ += diff
            else:
                if abs(diff) >= velocity:
                    self.robot_position_ -= velocity
                else:
                    self.robot_position_ -= abs(diff)

            self.get_logger().info("Robot position: " + str(self.robot_position_))
            feedback.current_position = self.robot_position_
            goal_handle.publish_feedback(feedback)

            time.sleep(1.0)

def main(args=None):
    rclpy.init(args=args)
    node = MoveRobotServerNode()
    rclpy.spin(node, MultiThreadedExecutor())
    rclpy.shutdown()


if __name__ == "__main__":
    main()


 # Levantar robot A
 # ros2 run lifecycle_py move_robot_server --ros-args -r __node:=move_robot_server_a -p robot_name:=A
 # Pasar estado configure
 #  ros2 lifecycle set /move_robot_server_a configure
 # Activar 
 # ros2 lifecycle set /move_robot_server_a activate

 # Levantar robot B
 # ros2 run lifecycle_py move_robot_server --ros-args -r __node:=move_robot_server_b -p robot_name:=B
 # Pasar estado configure
 #  ros2 lifecycle set /move_robot_server_b configure
 # Activar 
 # ros2 lifecycle set /move_robot_server_b activate
